clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,242 @@
+			WithError(err).
+			Error("Failed to launch etcd process")
+		return err
+	}
+	etcdPid := etcdCmd.Process.Pid
+	log.WithField("etcdPID", etcdPid).
+		Info("Local etcd server process started")
+
+	// Defer etcd process cleanup
+	defer func() {
+		log := log.WithField("etcdPID", etcdPid)
+		log.Debug("Cleaning up etcd process")
+		// Send the process a SIGTERM. SIGTERM is the "gentle" shutdown signal, and etcd should close down it's resources
+		// cleanly and then exit.
+		log.Info("Sending SIGTERM signal to etcd process")
+		err := etcdCmd.Process.Signal(syscall.SIGTERM)
+		if err != nil {
+			log.WithError(err).
+				Error("Failed to send SIGTERM signal to etcd process")
+			returnErr = err
+		}
+
+		// Wait for the etcd process to finish, and cleanup resources.
+		log.Info("Waiting for etcd process to exit")
+		err = etcdCmd.Wait()
+		if err != nil {
+			if exitError, ok := err.(*exec.ExitError); ok {
+				if exitError.ExitCode() == -1 {
+					// We SIGTERMed the etcd process, so a nonzero exit code is expected.
+					// Check the context as a last sanity check
+					if ctx.Err() != nil {
+						// Don't log the error itself here, if the context is timed out it'll be cancelled, so the error
+						// will just say "context cancelled" and not be useful — and possibly even misleading. It's
+						// possible that the timeout expires at the moment between etcd exiting normally and this check,
+						// which would report a false error. That's very unlikely, so we don't worry about it here.
+						log.WithField("timeout", timeout).
+							Error("etcd exited, but our context has expired. etcd may have been terminated due to timeout. Consider increasing the value of the timeout using the --timeout flag or CILIUM_TIMEOUT environment variable.")
+						// If we're not already returning an error, return an error here
+						if returnErr == nil {
+							returnErr = ctx.Err()
+						}
+						return
+					}
+					// This is the "good state", the context hasn't expired, the etcd process has exited, and we're
+					// okay with a nonzero exit code because we exited it with a SIGTERM.
+					log.Info("etcd process exited")
+					return
+				}
+				log.WithError(err).
+					WithField("etcdExitCode", exitError.ExitCode()).
+					Error("etcd process exited improperly")
+				// If we're not already returning an error, return an error here
+				if returnErr == nil {
+					returnErr = err
+				}
+				return
+			} else {
+				// Some other kind of error
+				log.WithError(err).
+					Error("Failed to wait on etcd process finishing")
+				// If we're not already returning an error, return an error here
+				if returnErr == nil {
+					returnErr = err
+				}
+				return
+			}
+		}
+		log.Info("etcd process exited")
+	}()
+
+	// With the etcd server process launched, we need to construct an etcd client
+	config := clientv3.Config{
+		Context:   ctx,
+		Endpoints: []string{loopbackEndpoint},
+	}
+	log.WithField("etcdClientConfig", fmt.Sprintf("%+v", config)).
```

Comment:
My thought was that we can print the whole config in case there was some change in defaults down the road that caused a headache. This only turns up in debug mode, or if the client creation fails anyway.

Reply:
I guess that should not happen (as it would be an etcd breaking change), but I'm not opposed to keeping the log as of now if you prefer so.